### PR TITLE
feat: adds Moonraker Job Queue Support

### PR DESCRIPTION
**Pull request by bwnance**
Added support for Moonraker's job queue system (https://moonraker.readthedocs.io/en/latest/web_api/#job-queue-apis)

image

image

image

Signed-off-by: Brandon Nance bwnance@gmail.com

### CODE REVIEW

File: src/store/files/mutations.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import Vue from 'vue'
 import { MutationTree } from 'vuex'
 import mergeFileUpdate from '@/util/merge-file-update'
-import { Files, FilesState, FileUpdate, AppFile, AppFileWithMeta, FileRoot } from './types'
+import { Files, FilesState, FileUpdate, AppFile, AppFileWithMeta, FileRoot, QueueJob } from './types'
```
Comment (pedrolamas) [inline src/store/files/mutations.ts:4]: The `QueueJob` import here is not in use, please remove.

File: src/router/index.ts [hunk 2]
```diff
@@ -54,6 +55,12 @@ const routes: Array<RouteConfig> = [
     component: History,
     beforeEnter: ifAuthenticated
   },
+  {
+    path: '/queue',
+    name: 'Qeuue',
```
Comment (pedrolamas) [inline src/router/index.ts:60]: Typo here in `Qeuue`

**Review by pedrolamas — CHANGES_REQUESTED** [review]
Thank your for your awesome work here!!

There are some minor inconsistency in terms of empty lines that have either been added or are missing (`socketActions.ts` is a good example of that), can you check on those please?

The rest of the code seems to be working fine, once you clear these minor issues I think we will be able to merge this in!

**Review by matmen — CHANGES_REQUESTED** [review]
I would love to get this landed in 1.18! However, there are a couple things that don't seem right:
1. The job queue card can't be disabled or moved in the layout editor:
image
2. I don't see a reason for having a job explorer above the job queue card on the Queue page. However, if it is required, it definitely needs a separator between it and the queue:
image
3. I'm not sure if that new page is necessary at all? Seems like the same settings/actions are also available from the card in the dashboard?

**Comment by matmen** [comment]
@bwnance Hey, just checking in, any updates on this or feedback to my comment? Thanks!

**Comment by bwnance** [comment]
@matmen I'm totally fine to remove that extra page and only have it in the dashboard. I thought it made sense to duplicate data in the dash to pages, but with the minimal amount of extra info, I think it's fine. Especially since we can queue files from the context menu.

**Comment by bwnance** [comment]
@matmen ok so i'll remove the Queue page. is it ok to leave it in but disabled? it's feasible that there will be some functionality added later that would warrant a dedicated queue page, and would save someone in the future the work

**Comment by matmen** [comment]
Not sure about leaving unreferenced files in the codebase, considering they are always stored in Git. Will need to see what Pedro's opinion on this is.
I can't seem to disable the queue card in the layout view (the checkbox is there, but unchecking it does nothing).

**Comment by bwnance** [comment]
@matmen ah sorry. didn't test in-depth enough. will resolve that. and yeah, up to @pedrolamas on leaving those files in. I have no problem deleting if it's an issue

**Comment by bwnance** [comment]
@matmen resolved in [f353aae](https://github.com/fluidd-core/fluidd/pull/448/commits/f353aae8ebe691119ce7f9bc2fe63bb207b9dfde)

File: src/mixins/files.ts [hunk 3]
```diff
@@ -150,6 +150,43 @@ export default class FilesMixin extends Vue {
       })
   }
 
+  enqueueFile (filename: string, path: string) {
```
Comment (matmen) [inline src/mixins/files.ts:153]: Is there a reason for this being a REST call instead of jsonrpc?

File: src/mixins/files.ts [hunk 4]
```diff
@@ -150,6 +150,43 @@ export default class FilesMixin extends Vue {
...
+    const data = {
+      filenames: filenames
+    }
+    return httpClient
+      .post(
+        this.apiUrl + '/server/job_queue/job',
+        data
+      )
+      .then((response) => {
+        this.updateQueue()
```
Comment (matmen) [inline src/mixins/files.ts:166]: The updated queue data is already returned by this POST, I don't see a need for calling this here

File: src/mixins/files.ts [hunk 5]
```diff
@@ -150,6 +150,43 @@ export default class FilesMixin extends Vue {
...
+        return response
+      })
+      .catch(e => {
+        return e
+      })
+  }
+
+  updateQueue () {
```
Comment (matmen) [inline src/mixins/files.ts:174]: Shouldn't the updates come in over the WS? Specifically `socket/notifyJobQueueChanged`

**Comment by pedrolamas** [comment]
> @matmen ah sorry. didn't test in-depth enough. will resolve that. and yeah, up to @pedrolamas on leaving those files in. I have no problem deleting if it's an issue

Ideally, I prefer not to leave dead (unused) code in the repository (that's what I use separate branches for! 😁)

**Comment by bwnance** [comment]
> > @matmen ah sorry. didn't test in-depth enough. will resolve that. and yeah, up to @pedrolamas on leaving those files in. I have no problem deleting if it's an issue
> 
> Ideally, I prefer not to leave dead (unused) code in the repository (that's what I use separate branches for! 😁)

Ok! Will remove dead code

File: src/components/widgets/queue/QueueCard.vue [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        <v-icon
+          small
+          left
+        >
+          $pause
+        </v-icon>
+        <span>{{ $t('app.general.btn.pause') }}</span>
+      </app-btn>
+      <app-btn
+        v-else-if="queueStatus == 'paused'"
```
[the remark this answers is not in the document]
Comment (pedrolamas) [reply src/components/widgets/queue/QueueCard.vue]: Fixed!

File: src/mixins/files.ts (hunk 4 above)
Comment (pedrolamas) [reply src/mixins/files.ts]: Removed!

File: src/mixins/files.ts (hunk 5 above)
Comment (pedrolamas) [reply src/mixins/files.ts]: Changed!

**Comment by pedrolamas** [comment]
I've been refactoring and updating this code, in the hope to get it merged on the v1.23 release!

Here are some screenshots of the current state:

image

image

image

**Comment by bwnance** [comment]
@pedrolamas thank you so much for taking this one over!! I haven't had much time to work on it, and the use-case i had for it is no longer with me, haha.

**Comment by pedrolamas** [comment]
No worries, to be honest I don't think I have a personal use either, but I see a lot of value in your work for those who do actually use the Job Queue, so I want to make sure we can get this merged and released ASAP!

**Review by pedrolamas — APPROVED** [review]

**Comment by pedrolamas** [comment]
Merry Christmas everyone!! 🎅

I now consider this work complete and ready for PR review! 😁

**Comment by andresandia** [comment]
Thanks for making queue available on fluidd, I'm working on a belt printer project, is there a way to set "resume" print automatically?, since the belt printer will release parts, we will need to automatically start the new print after part is released, idea is not to babysit the printer nor press resume button in order to start new print, thanks in advance for your reply.

**Comment by pedrolamas** [comment]
@andresandia you should only need to click "resume" once, to start the first print. After that, it will carry on picking items from the queue to print!

**Pull request merged**